server/mediaserver.go

```diff
@@ -162,9 +164,15 @@ func createRTMPStreamIDHandler(s *LivepeerServer) func(url *url.URL) (strmID str
 
 		// Ensure there's no concurrent StreamID with the same name
 		s.connectionLock.RLock()
-		defer s.connectionLock.RUnlock()
 		if _, exists := s.rtmpConnections[mid]; exists {
 			glog.Error("Manifest already exists ", mid)
+			s.connectionLock.RUnlock()
+			return ""
+		}
+		s.connectionLock.RUnlock()
+
+		if !authenticateStream(string(mid)) {
```

Comment:
Let's authenticate first thing, before acquiring the lock.

I'm thinking we should pass in the full URL rather than the ManifestID. That way the auth endpoint can inspect the entire host + path, and enforce their own rules about path construction.

The thought was to have the webhook URL return the manifest ID. We should also structure the response to leave room for further customization, eg transcoding options for the stream.

```
GET https://host/webhook/endpoint
    Request Body: { "url" : <url> }
    Response Body: { "manifestID" : <manifestID> }
```
We could also do a POST here since a GET with a body is a bit unusual.

````golang
func createRTMPStreamIDHandler(s *LivepeerServer) func(url *url.URL) (strmID string) {     
     return func(url *url.URL) (strmID string) {     
        // Create a ManifestID
        // If an auth webhook is defined, get the mid from the webhook
        // Else if manifestID is passed in, use that one
        // Else create one
        var mid string 
        if WebhookURL != "" {
          mid, err = authenticateStream(url)
          if err != nil {
            return ""
          }
        } else {
          mid = parseManifestID(...)
        }
```

Reply:
> Let's authenticate first thing, before acquiring the lock.

Previous check will allow us to quit early in case of same Manifest ids, so why not?

> The thought was to have the webhook URL return the manifest ID

I really don't remember discussing such thing. Can you point me to such discussion?

>We should also structure the response to leave room for further customization,

This is starting to be something other from what we're discussed.

Reply:
> Previous check will allow us to quit early in case of same Manifest ids

If the manifest ID is defined by the webhook then the webhook needs to come before.  Also avoids acquiring the lock which also helps.

> I really don't remember discussing such thing. Can you point me to such discussion?

Yeah, we had a lot of discussions around this and some details may have been lost. This was mentioned here https://hackmd.io/n0Lq1l6_SJCe0VnVHf4G-A#Webhook-Flow which was outlined as part of the discussion here : 
https://docs.google.com/a/livepeer.org/document/d/1ThTyFeTjTfnWWj-HeNCMnkglD7Vf-9YvA71fB9dKVQs/edit?disco=AAAACet6oNM

> This is starting to be something other from what we're discussed.

How so? The end result is the same, we aren't changing the scope of the work here. But tweaking the webhook format makes the construct much more versatile.

Once we merge this, it effectively becomes part of the public API, and it's harder to make breaking changes (even internally), so it helps a lot to leave a little room for expansion.

Reply:
> https://hackmd.io/n0Lq1l6_SJCe0VnVHf4G-A#Webhook-Flow

missed this one

> was outlined as part of the discussion here :
https://docs.google.com/a/livepeer.org/document/d/1ThTyFeTjTfnWWj-HeNCMnkglD7Vf-9YvA71fB9dKVQs/edit?disco=AAAACet6oNM

Here is no mention of webhook defining manifest ids. 

> How so? The end result is the same, we aren't changing the scope of the work here.

Webhook's server should be more complicated now, plus our logic gets more complicated - what to do if webhook returned manifest id that we already have? Or if it is invalid manifest id?

Reply:
> Here is no mention of webhook defining manifest ids.

There seems to have been quite a bit lost "in translation" when migrating from the [old doc](https://docs.google.com/document/d/16Z0xvIYe8FNcKJ76HCVZ8JJ8xfHytr9te5GRNrE3XhU/edit) to the [new one](https://docs.google.com/document/d/1ThTyFeTjTfnWWj-HeNCMnkglD7Vf-9YvA71fB9dKVQs/edit#).

> Alternative Implementation - Webhook:
> Start up node with a `-streamAuthWebhook http://host/streamauth`
> RTMP stream comes in
> Node calls the streamAuthWebhook with the RTMP URL, and  other parameters?
> Node receives manifest ID in response (disconnects incoming rtmp stream if webhook  response indicates an invalid stream)
> Stream is available for playback via the manifest ID

The behavior spec in the old doc also details how it would be desirable to have the input URL to be decoupled from the output URL to prevent input hijacking or guessing playback URLs given an input. Having the webhook return the manifest ID enables that.

Do you think that is not useful?

> Webhook's server should be more complicated now

Reading a JSON request and replying with JSON is standard, as far as APIs go.

> plus our logic gets more complicated

Our logic stays the same. The only thing that changes is the webhook gets called first to set the manifest id, if necessary.

A webhook is a large feature as far as API surface area goes, and we want to extract as much utility from it as possible. Right now its utility is limited.

Reply:
I've added POST API, with only one addition - if webhook returns 200 with empty body - it is enough to pass authentication. This will simplify creation of simple webhooks.